lib/components/user/phone-number-editor.tsx

```diff
@@ -0,0 +1,332 @@
+
+/**
+ * Sub-component that handles phone number and validation code editing and validation intricacies.
+ */
+class PhoneNumberEditor extends Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+
+    const { initialPhoneNumber } = props
+    this.state = {
+      // If true, phone number is being edited.
+      // For new users, render component in editing state.
+      isEditing: isBlank(initialPhoneNumber),
+
+      // Holds the new phone number (+15555550123 format) entered by the user
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
+  }
+
+  _handleEditNumber = () => {
+    this.setState({
+      isEditing: true,
+      newPhoneNumber: ''
+    })
+  }
+
+  _handleNewPhoneNumberChange = (newPhoneNumber: string) => {
+    this.setState({
+      newPhoneNumber
+    })
+  }
+
+  _handleCancelEditNumber = () => {
+    this.setState({
+      isEditing: false
+    })
+  }
+
+  _handlePhoneNumberKeyDown = (e: KeyboardEvent) => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the phone number field,
+      // prevent form submission and request the code.
+      e.preventDefault()
+      this._handleRequestCode()
+    }
+  }
+
+  _handleValidationCodeKeyDown = (e: KeyboardEvent) => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the validation code field,
+      // prevent form submission and send the validation code.
+      e.preventDefault()
+      this._handleSubmitCode()
+    }
+  }
+
+  /**
+   * Send phone verification request with the entered values.
+   */
+  _handleRequestCode = () => {
+    const { initialPhoneNumber, initialPhoneNumberVerified, onRequestCode } =
+      this.props
+    const { newPhoneNumber } = this.state
+
+    this._handleCancelEditNumber()
+
+    // Send the SMS request if one of these conditions apply:
+    // - the user entered a valid phone number different than their current verified number,
+    // - the user clicks 'Request new code' for an already pending number
+    //   (they could have refreshed the page in between).
+    let submittedNumber
+
+    if (
+      newPhoneNumber &&
+      isPossiblePhoneNumber(newPhoneNumber) &&
+      !(newPhoneNumber === initialPhoneNumber && initialPhoneNumberVerified)
+    ) {
+      submittedNumber = newPhoneNumber
+    } else if (this._isPhoneNumberPending()) {
+      submittedNumber = initialPhoneNumber
+    }
+
+    if (submittedNumber) {
+      onRequestCode(submittedNumber)
+    }
+  }
+
+  _handleSubmitCode = () => {
+    const { errors, onSubmitCode, values } = this.props
+    const { validationCode } = values
+
+    if (!errors.validationCode) {
+      onSubmitCode(validationCode)
+    }
+  }
+
+  _isPhoneNumberPending = () => {
+    const { initialPhoneNumber, initialPhoneNumberVerified } = this.props
+    return !isBlank(initialPhoneNumber) && !initialPhoneNumberVerified
```

Comment:
I'm noticing that the component briefly flashes validated before switching to pending. I'm not sure if this was introduced by these changes, but it seems to be related to this check

Reply:
Correct, that is an old issue.

Reply:
This is fixed in https://github.com/opentripplanner/otp-react-redux/pull/774/commits/e2936f327849821f02bbe10751910ac0808a1483.